### PR TITLE
Provide callback to modify logged values

### PR DESCRIPTION
**Pull request by camertron**
Let's you modify the values of logged attributes via the block passed to `automatically_log_changes` class method (which has replaced the `after_initialize` callback setup). This is useful when you want to avoid serializing a complex Ruby object like a `CarrierWave::Uploader`. Here's an example:

```ruby
class Vehicle < ActiveRecord::Base
  # an example carrierwave uploader attached to the 'photo' attribute
  mount_uploader :photo, MyApp::VehiclePhotoUploader

  include ChangesAreLogged

  automatically_log_changes do |attribute, old_value, new_value|
    if attribute == 'photo'
      [old_value, new_value.sanitized_file.original_filename]
    end
  end
end
```

Now `photo` won't get YAML serialized and written into the `changes_logged` column in your `change_logs` table.

@lumoslabs/platform @lumoslabs/rails

### CODE REVIEW

File: lib/changes_are_logged.rb [hunk 1]
```diff
@@ -22,16 +22,26 @@ def log_it
         @changes_logged = {}
         save_change_log
       elsif self.changed? || !@change_comments.blank?
-        @changes_logged = self.changes
+        exclude_list = self.class.log_changes_options.fetch(:exclude, []).map(&:to_s)
```
Comment (zvkemp) [inline lib/changes_are_logged.rb]: propose using `:except`, which is a more-or-less standard option on a number of built-in rails methods

File: lib/changes_are_logged.rb [hunk 2]
```diff
@@ -22,16 +22,26 @@ def log_it
...
+        include_list = self.class.columns.map(&:name) - exclude_list
+
+        @changes_logged = self.changes.each_with_object({}) do |(attribute, value), ret|
```
Comment (zvkemp) [inline lib/changes_are_logged.rb]: `self.` is redundant

File: lib/changes_are_logged.rb [hunk 3]
```diff
@@ -22,16 +22,26 @@ def log_it
...
+          ret[attribute] = if include_list.include?(attribute.to_s)
+            value
+          else
+            [nil, 'Attribute changed, but value has been excluded.']
```
Comment (zvkemp) [inline lib/changes_are_logged.rb]: I would expect this condition not to return a value at all — maybe there should be a `:mask` option that indicates a change was made, but the values aren't revealed in the log?

File: lib/changes_are_logged.rb [hunk 4]
```diff
@@ -43,15 +53,21 @@ def save_change_log
...
+  end
+
+  module ClassMethods
+    attr_reader :log_changes_options
 
-    def automatically_log_changes
-      @log_changes = true
+    def automatically_log_changes(options = {})
+      after_initialize -> { @log_changes = true }
+      @log_changes_options = options
```
Comment (zvkemp) [inline lib/changes_are_logged.rb]: this won't be inherited for STI models

**Review by zvkemp — CHANGES_REQUESTED** [review]

File: lib/changes_are_logged.rb (hunk 4 above)
Comment (camertron) [reply lib/changes_are_logged.rb]: Is that a problem?

File: lib/changes_are_logged.rb (hunk 3 above)
Comment (camertron) [reply lib/changes_are_logged.rb]: Ok, `:mask` makes sense. What do you think of `:filter`?

File: lib/changes_are_logged.rb (hunk 2 above)
Comment (camertron) [reply lib/changes_are_logged.rb]: Good catch, copied/pasted from legacy.

File: lib/changes_are_logged.rb (hunk 3 above)
Comment (zvkemp) [reply lib/changes_are_logged.rb]: `:filter` sounds good

File: lib/changes_are_logged.rb (hunk 4 above)
Comment (zvkemp) [reply lib/changes_are_logged.rb]: IMO yes, since the `after_initialize` block *will* be inherited, but the options won't be.

**Review by zvkemp — APPROVED** [review]

**Pull request merged**